### ISSUE TITLE
roles vars vs host vars

### ISSUE THREAD

**Issue by hellt** [srl-labs/intent-based-ansible-lab#5]
@wdesmedt I wanted to check why did you decide to add host variables in each role's vars section. Why don't you put host vars into the inventory, making it a single place where host_vars are defined.

Imagine you grow your fabric from 4 leaves to 6 leaves. With the current approach you would have to add clab-xxx-l5 and clab-xxx-l6 vars to all the roles, instead of just adding the relevant nodes into the inventory.

**Comment by wdesmedt** [comment]
@hellt The 'intention' was to keep the variables specific to the role with that role. So to configure interfaces, you go to the vars of the interface role, same for network-instance. You could also put this in the host_vars and group_vars of the inventory but it's a matter of opinion which is best. I also wanted to be independent of an external inventory like netbox for intents.

**Comment by hellt** [comment]
I don't think I ever saw host-specific variables specified in the role's vars; it just doesn't fit in my view. The whole point of the inventory file is to have a single source of truth for the configuration management generation. By moving this to a role, you have many places where inventory is "redefined".

The role vars, afair, are predominantly used to store the defaults the role can use if unset from other sources. E.g. interface speed might be set to 1G if not specified in the host/group/playbook.

You can imagine how frustrating it would be to change all the roles if we, say, rename the lab name from 4l2s to 6l2s.
I'll leave this to you, but maybe it is worth checking with other projects and get some references backing this design decision.

**Comment by wdesmedt** [comment]
One of the reasons to have a dedicated role to load intent (`utils/load_intent`) that defines host and group vars - mimicking host_vars and group_vars in the ansible inventory - was the behaviour of Ansible with overlapping vars in host_vars and group_vars, e.g.:

host_vars: 

```yaml
network_instance:
  default:
    protocols:
      bgp:
        neighbor:
          192.168.0.0:
            peer_group: spines
          192.168.1.0:
            peer_group: spines
```

group_vars:

```yaml
network_instance:
  default:
    interfaces:
      - system0.0
      - ethernet-1/48.0
      - ethernet-1/49.0
    autonomous-system: "{{ asn }}"
    protocols:
      bgp:
        neighbor:
          192.168.255.101:
            peer_group: overlay
          192.168.255.102:
            peer_group: overlay
        ipv4:
          admin_state: enable
        group:
          overlay:
            peer_as: 100
            export_policy: pass-evpn
            import_policy: pass-evpn
            ipv4:
              admin_state: disable
            evpn:
              admin_state: enable
            local_as: 100
            local_address: "{{ router_id }}"
          spines:
            description: "Peer-group for spine neighbors"
            export_policy: lo-and-servers
            import_policy: pass-all
            peer_as: 65100
            ipv4:
              admin_state: enable
            epvn:
              admin_state: disable
```

The default behaviour of Ansible for duplicate var definition is controlled by the setting `hash_behaviour=replace` in `ansible.cfg`. Changing this to `merge`, which is required for our use-case, is not recommended as stated in the elaborate version of ansible.cfg: 

> *...[we] recommend relying on the `combine` filter and `var` lookups. 

Also, relying on a setting in ansible.cfg to control fundamental behavior of the playbook (building intent) is risky. 

So `host_vars` and `group_vars` in the inventory to store intent is not the way to store variables that are deeply nested objects and have both a host-specific and group-specific definition. Probably a single file with all the intents *outside* the inventory is a middle way forward.

**Comment by hellt** [comment]
either indeed making the separate inventory file with variables, or using a simpler variables structure with non overlapping fields between host vars and group vars

**Comment by wdesmedt** [comment]
addressed in PR #18

**Comment by wdesmedt** [comment]
fixed in dev branch

**Comment by wdesmedt** [comment]
Fix now in main #34
